Add LSP documentation (#263)
UMAprotocol/docs: docs/synthetic-tokens/long-short-pair.md

- Covered call options.
- [Range Tokens for treasury management](https://medium.com/uma-project/treasury-diversification-with-range-tokens-145d4b12614e).

> ## LSP Contract Deployment

Reviewer: This section seems unnecessary since there is a separate LSP deployment tutorial. Can you remove this section and just replace with adding one sentence to the previous section linking to the deployment tutorial?

Author: I agree with this. Including a sentence in the section above and removing.

### `create`

After an LSP contract has been deployed, tokens can be minted by calling the `create` function before the contract’s `expirationTimestamp`. The `create` function deposits collateral into the contract in exchange for an equal amount of long and short tokens based on the `collateralPerPair` parameter. The `collateralPerPair` parameter determines the amount of collateral that is required for each pair of long and short tokens. 

![](/docs/lsp-tokens/lsp_create.png)

> To create tokens, first approve the token minting contract to transfer the collateral currency on your behalf. Then input the number of tokens you would like to create into the Write Contract tab of the LSP contract in Etherscan. If the collateral is WETH and the `collateralPerPair` is set to 1, each long and short token minted would require 1 WETH as collateral.

Reviewer: Recommend adding one more example that is less intuitive - if collateralPerPair is 0.25 etc.

Author: Updated to include an example with .25 WETH.

After tokens are minted, to confirm that tokens have been issued by the contract, go to the Read Contract tab of the LSP contract in Etherscan and call the `getPositionTokens` function with the address you minted the tokens from. This will return the number of long and short tokens which will be an equal amount after you mint. The short and long tokens received represent a fully collateralized and risk-neutral position. A minter only receives long or short exposure by selling one of the tokens while a buyer receives exposure simply by purchasing one of the tokens.

### `redeem`

At any point pre-expiry, anyone can call `redeem` with a pair of long and short tokens and receive an amount of collateral determined by `collateralPerPair`. The total amount of collateral received from the contract is determined by `collateralPerPair` * `tokensToRedeem`.

> Regardless of the value of the long and short tokens when redeemed, the summed value of the two tokens will always be worth the total amount of collateral used to mint. The contract burns the tokens and returns collateral proportional to the amount that the token sponsor has deposited to the contract. The caller does not need to approve this contract to transfer any number of redeemed tokens since long and short tokens are burned, rather than transferred.

Reviewer: ```suggestion
Regardless of the value of the long and short tokens when redeemed, the summed value of the two tokens will always be worth the total amount of collateral used to mint (collateralPerPair). The contract burns the tokens and returns collateral proportional to the amount that the token sponsor has deposited to the contract. The caller does not need to approve this contract to transfer any number of redeemed tokens since long and short tokens are burned, rather than transferred.
```

![](/docs/lsp-tokens/lsp_redeem.png)

### `expire`

When the current timestamp is later than the `expirationTimestamp` parameter, token holders are unable to settle their tokens for collateral until a price has been received by the Optimistic Oracle. The `settle` function will revert until the `expire` function is called once by anyone. 

The `expire` function does not take any parameters and requests a price from the Optimistic Oracle for the LSP contract's `priceIdentifier`, `expirationTimestamp`, `customAncillaryData`, `collateralToken`, and `prepaidProposerReward`. At this time, a proposal bond can be set associated with the price request using `setBond` along with a custom liveness value for the request using `setCustomLiveness`. Liveness is the amount of time a proposal must wait before being resolved.

![](/docs/lsp-tokens/lsp_expire.png)

> ### `settle`

Reviewer: One other thing I'd add here is the concept of the contract's state. This can be read for the LSP with `contractState`.

Enum values for your context:
- Value of 0: Open (price has not been requested)
- Value of 1: `ExpiredPriceRequested` (one the price has been requested). Small nuance here is that it will still be in this state until the first person calls `settle`, even if a price has been returned to the OO. This is because the LSP has no context that the price has been resolved by the OO until requested successfully.
- Value of 2: `ExpiredPriceReceived` (price was available from OO and first settle call has been performed successfully)

Author: Added a section to include.

> Once a price request exists, Proposers respond by referencing off-chain price feeds and submitting parameters for the `priceIdentifier`, `timestamp`, `ancillaryData`, and `proposedPrice` through the Optimistic Oracle contract. In return, Proposers receive a pre-defined proposal reward set by the Requestor. To propose prices, the Proposer is required to stake a proposal bond. If the price information they provided is disputed and deemed incorrect, the Proposer will lose their bond.

Reviewer: Recommend saying something about bond amounts being custom per LSP and can be set on deployment. Same for liveness below.

Author: I agree. I updated but let me know if you want me to make any changes on how I implemented the change.

> Disputers can refute a price submitted by a Proposer within the proposal liveness period by referencing their own off-chain price feeds. The proposal liveness period is a pre-defined amount of time a proposal can be disputed before the Requestor receives the price of the asset. If Disputers do not refute the price submitted by the Proposer within the proposal liveness period, the price is sent to the Requestor. If a proposal is disputed, the price will be submitted to UMA’s DVM and resolved after 48-hours.

Reviewer: ```suggestion
Disputers can refute a price submitted by a Proposer within the proposal liveness period by referencing their own off-chain price feeds. The proposal liveness period is a pre-defined amount of time a proposal can be disputed before the Requestor receives the price of the asset. If Disputers do not refute the price submitted by the Proposer within the proposal liveness period, the price is treated as correct and can now be read from the Optimistic Oracle. If a proposal is disputed, the price will be escalated to UMA’s Decentralized Verification Mechanism (DVM) and resolved after a 48-96 voting period.
```

Author: Make an update but want to confirm, shouldn't it be...after a 48-96 hour voting period?

Reviewer: Yes correct. Isn't that what you have now?

> ![](/docs/lsp-tokens/lsp_settle.png)

Reviewer: Great graphic, one small recommendation would be to add an arrow from the LSP contract to the OO for the price request, and an arrow from the DVM to the OO for a disputed price response.

Also small text change, but the price isn't actually sent from the OO to the LSP automatically. The LSP needs to read it from the OO.

Author: Updated the graphic to included the changes. Let me know if there are any other ways you can think of improving it.

To propose a price through Etherscan, use the Write Contract tab in the Optimistic Oracle contract to find proposePrice. For reference, [here](https://github.com/UMAprotocol/protocol/tree/master/packages/core/networks) is a list of contract addresses for each network that can be used to find the Optimistic Oracle contract. The parameters for proposing a price are:

- `Requestor`: sender of the initial price request

> - `Identifier`: price identifier to identify the existing request

Reviewer: ```suggestion
- `Identifier`: price identifier to identify the existing request in bytes32 format.
```

- `timestamp`: timestamp to identify the existing request
- `ancillaryData`: ancillary data of the price being requested

> - `proposedPrice`: price being proposed

Reviewer: ```suggestion
- `proposedPrice`: price being proposed scaled to 18 decimals
```

> After a price has been proposed, the same parameters can then be used with the `hasPrice` and `getRequest` functions. The `hasPrice` queries whether a request has been resolved. If `hasPrice` is equal to true, values will be returned for `expiryPrice` and `expiryPercentLong` in the Read Contract tab for the LSP contract and the `settle` function can be called. The `getRequest` function gets the current data structure containing all information about a price request.

Reviewer: Also same nuance as above, but 
```suggestion
After a price has been proposed, the same parameters can then be used with the `hasPrice` and `getRequest` functions. The `hasPrice` queries whether a request has been resolved. If `hasPrice` is true, a price is available and the `settle` function can be called on the LSP. Once the first `settle` call is executed, the LSP's state will update with the resolved price. At this point, the price and its settlement outcome can be read with the `expiryPrice` and `expiryPercentLong` functions. The `getRequest` function gets the current data structure containing all information about a price request.
```

> The `expiryPrice` is the value returned by the Optimistic Oracle for the contract's `expirationTimestamp` which is used by the `financialProductLibrary` to calculate `expiryPercentLong` and determine the redemption rate between long and short tokens. `ExpiryPercentLong` is a number between 0 and 1, where 0 sends all collateral to the short tokens and 1 sends all collateral to the long tokens. The collateral returned is the sum of the two payouts and both the long and short tokens are burned. A simple calculation for how to use these values to predict payouts is shown below.

Reviewer: ```suggestion
The `expiryPrice` is the value returned by the Optimistic Oracle for the contract's `expirationTimestamp` which is used by the `financialProductLibrary` to calculate `expiryPercentLong` and determine the redemption rate between long and short tokens. `ExpiryPercentLong` is a number between 0 and 1, where 0 assigns all collateral to the short tokens and 1 assigns all collateral to the long tokens. The collateral returned is the sum of the two payouts and both the long and short tokens are burned. A simple calculation for how to use these values to predict payouts is shown below.
```
